File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]
# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
#library(caTools)
#set.seed(123)
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#training_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting linear regression model to the dataset.
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

# Fititng polynomial regression model to the dataset.
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ . ,
              data = dataset)

# Visualizing the linear regression results.
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level,y = dataset$Salary), colour='Red') +
  geom_line(aes(x = dataset$Level ,y = predict(lin_reg , newdata = dataset) , colour = 'Blue')) +
  xlab('Position Level') +
  ylab('Salary') +
  ggtitle('Truth or Bluff (Linear Regression Results)')
  

# Visualizing the Polynomial regression results.

ggplot() +
  geom_point(aes(x = dataset$Level,y = dataset$Salary), colour='Red') +
  geom_line(aes(x = dataset$Level ,y = predict(poly_reg , newdata = dataset) , colour = 'Blue')) +
  xlab('Position Level') +
  ylab('Salary') +
  ggtitle('Truth or Bluff (Polynomial Regression Results)')

# Predicting a new result with Linear Regression
y_pred = predict(lin_reg , data.frame(Level = 6.5))

# Predicting a new result with Linear Regression
y_pred_Poly = predict(poly_reg , data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jan 26 11:19:14 2018

@author: ak901t
"""

# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
# from sklearn.cross_validation import train_test_split
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting Linear regression to the dataset.
from sklearn.linear_model import LinearRegression
linearRegressor1 = LinearRegression()
linearRegressor1.fit(X,y);

# Fitting polynomial regression to the dataset.
from sklearn.preprocessing import PolynomialFeatures
polynomialRegressor = PolynomialFeatures(degree=4)
X_poly = polynomialRegressor.fit_transform(X)
linearRegressor2 = LinearRegression()
linearRegressor2.fit(X_poly , y)

# Visualising the linear regression results
plt.scatter(X, y, c='Red')
plt.plot(X, linearRegressor1.predict(X), c='Blue')
plt.title('Truth or Bluff (Linear Regression Results)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Visualising the polynomial regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid) , 1)
plt.scatter(X, y, c='Red')
plt.plot(X_grid, linearRegressor2.predict(polynomialRegressor.fit_transform(X_grid)), c='Blue')
plt.title('Truth or Bluff (Polynomial Regression Results)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with Linear Regression.
linearRegressor1.predict(6.5)

# Predicting a new result with Polynomial Regression.
linearRegressor2.predict(polynomialRegressor.fit_transform(6.5))